compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java

```diff
@@ -1963,8 +1965,10 @@ public void visit(BLangRecordLiteral.BLangMapLiteral mapLiteral) {
     public void visit(BLangRecordLiteral.BLangStructLiteral structLiteral) {
         SymbolEnv symbolEnv = env.createClone();
         BLangFunction enclInvokable = (BLangFunction) symbolEnv.enclInvokable;
-        structLiteral.enclMapSymbols = collectClosureMapSymbols(symbolEnv, enclInvokable, false);
-
+        if (structLiteral.getBType().getKind() == TypeKind.RECORD) {
+            ((BRecordType) structLiteral.getBType()).enclMapSymbols =
+                                                     collectClosureMapSymbols(symbolEnv, enclInvokable, false);
+        }
```

Comment:
Why can't we do this when visiting the type?

Reply:
Updated

Reply:
Has the PR been updated?

Reply:
I forgot to update the comment. The type definitions are resolving before visiting functions, so all the enclosing map symbols cannot be captured at that stage.

Reply:
That should be the case only for module-level definitions and anon records, right? For local anon records, they should be visited when the function is being visited?

Reply:
Doesn't this also mean `enclMapSymbols` isn't properly set for local anon records if a value isn't created using that record typedesc?

Reply:
Ping.

Reply:
Yes, When traversing a function, local record-type nodes are visited. But the problem is when visiting a record-type node, it is not known which enclosure map symbols are necessary for a specific record literal.
```ballerina
function value() {
    record{int x;} m;
    int z = 10;
    var x = function() {
            int k = z;
            var y = function() {
                m = {x:k+z};
            };
    };
}
```
Consider the above example, When visiting m, cannot identify what enclosure map symbols are needed for the record literal.

Reply:
Do you mean mapping constructor? Is `enclMapSymbols` relevant to the type or the value? If it is the latter, moving it to the type is wrong.

Reply:
Updated with https://github.com/ballerina-platform/ballerina-lang/pull/39101/commits/c0ef0a550116ffe5b0a0eaf59a6177e39afbbaaf